### PR TITLE
Feature/557 template to generate protocols from types

### PR DESCRIPTION
**Pull request by doozMen**
I would still need some help with the tests. But this adds:

Template for generating protocols
Example of use

### CODE REVIEW

File: Templates/Templates/AutoGenerateProtocol.stencil [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+//: - initializers marked with annotation // sourcery:includeInitInProtocol
+//: - of the above it does not add it if  // sourcery:skipProtocol
+//: ---
+
+{% for type in types.implementing.AutoGenerateProtocol %}
+// MARK: - {{ type.name }} AutoGenerated protocol
+protocol {{ type.name }}Protocol: AutoMockable {
+    {#loop over all variables from this type and not from the super class#}
+    {% for variable in type.variables where variable.readAccess != "private" %}
+    {%if not method.annotations.skipProtocol %}var {{variable.name}}: {{variable.typeName}} { {%if variable.isMutable %} get set {% else %} get {% endif %} }{% endif %}
```
[the remark this answers is not in the document]
Comment (ilyapuchka) [reply Templates/Templates/AutoGenerateProtocol.stencil]: And you can simplify by checking for mutability only to add `set` so you will not need `else` branch

File: Templates/Templates/AutoGenerateProtocol.stencil [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    {% endfor %}
+
+    {#optionally add initializers when included in protocol#}
+    {% for method in type.methods where method.annotations.includeInitInProtocol %}
+    {{method.name}}
+    {% endfor %}
+
+    {#loop over all methods static class or normal#}
+    {% for method in type.methods where method.accessLevel != "private" and method.isInitializer == 0 %}
```
Comment (ilyapuchka) [inline Templates/Templates/AutoGenerateProtocol.stencil]: why `method.isInitializer == 0` and not just `not method.isInitializer`?

File: Templates/Templates/AutoGenerateProtocol.stencil [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+//: From all Types implementing this protocol Sourcery adds:
+//: - public/internal variables // private variables are ignored
+//: - public/internal methods (skips initializers)
+//: - initializers marked with annotation // sourcery:includeInitInProtocol
+//: - of the above it does not add it if  // sourcery:skipProtocol
+//: ---
+
+{% for type in types.implementing.AutoGenerateProtocol %}
+// MARK: - {{ type.name }} AutoGenerated protocol
+protocol {{ type.name }}Protocol: AutoMockable {
```
Comment (ilyapuchka) [inline Templates/Templates/AutoGenerateProtocol.stencil]: I guess your main intention for this template is to generate protocols for mocks, but I think this conformance should be something that should be added manually.

File: Templates/Templates/AutoGenerateProtocol.stencil [hunk 4]
```diff
@@ -0,0 +1,32 @@
+
+//: Do not change this code as it is autogenerated every time you build.
+//: You can change the code in `../StencilTemplatesForSourcery/Application/AutoGenerateProtocol
```
Comment (ilyapuchka) [inline Templates/Templates/AutoGenerateProtocol.stencil]: What is this path?

File: Templates/Templates/AutoGenerateProtocol.stencil [hunk 5]
```diff
@@ -0,0 +1,32 @@
+
+//: Do not change this code as it is autogenerated every time you build.
```
Comment (ilyapuchka) [inline Templates/Templates/AutoGenerateProtocol.stencil]: We already have warning, don't think there is need to state this in template too.

File: Templates/Templates/AutoGenerateProtocol.stencil [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+//: You can change the code in `../StencilTemplatesForSourcery/Application/AutoGenerateProtocol
+
+// MARK: - AutoGenerateProtocol
+//: From all Types implementing this protocol Sourcery adds:
```
Comment (ilyapuchka) [inline Templates/Templates/AutoGenerateProtocol.stencil]: Not sure if documentation comments are needed in the generated code. I've proposed a format for templates documentation here #531 but didn't touch on that. @krzysztofzablocki what do you think?

File: Templates/Tests/Expected/AutoGenerateProtocol.expected [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+// MARK: - AutoGenerateProtocol
+//: From all Types implementing this protocol Sourcery adds:
+//: - public/internal variables // private variables are ignored
+//: - public/internal methods (skips initializers)
+//: - initializers marked with annotation // sourcery:includeInitInProtocol
+//: - of the above it does not add it if  // sourcery:skipProtocol
+//: ---
+
+// MARK: - Car AutoGenerated protocol
+protocol CarProtocol: AutoMockable {
```
Comment (ilyapuchka) [inline Templates/Tests/Expected/AutoGenerateProtocol.expected]: you should probably change either expected type or type in source. And you should actually test this template in `TemplatesTests.swift`

File: Templates/Templates/AutoGenerateProtocol.stencil (hunk 6 above)
Comment (krzysztofzablocki) [reply Templates/Templates/AutoGenerateProtocol.stencil]: Might be useful for people unfamiliar with Sourcery, but if we do add docs to generated code then I'd be 👍 for using the format you described, that we could eventually actually parse and run some extra checks on the tool side

**Comment by doozMen** [comment]
@ilyapuchka thanks for the feedback. I will look into it this week.

**Pull request closed without merging**